Refactoring bsc fixes

jvegasbsc opened the pull request:
A couple of fixes needed to use BSC main data repository but that I think must be welcome for more people

* input_files now supports drs like input_dir, but is not mandatory
* get_start_end_year now support files with only one date at the end (i.e tas_200001.nc)

and a new exception raised when for any reason load_cubes does not load any cubes

mattiarighi commented:
Please fix the codacy issue and (if possible) circle/ci tests.

bouweandela commented on esmvaltool/preprocessor/_io.py:
```diff
@@ -42,6 +42,8 @@ def load_cubes(files, filename, constraints=None, callback=None):
     cubes = cubes.concatenate()
     for cube in cubes:
         cube.attributes['_filename'] = filename
+    if len(cubes) == 0:
```
Replace by a more Pythonic
```
if not cubes:
```

jvegasbsc replied:
Yes, I have changed it also moved it before the for

bouweandela commented:
I think the Python 2 tests can safely be ignored for now, see issue #273.

mattiarighi commented:
Ok, thanks.

@jvegasbsc please fix the codacy issues.

bouweandela commented on esmvaltool/preprocessor/_reformat.py:
```diff
@@ -29,9 +32,12 @@ def _checker(cube):
 
 def fix_file(filename, short_name, project, model, preproc_dir):
     """Fix errors that prevent loading or can not be fixed in the cube."""
+    final_dir = os.path.join(preproc_dir, project, model, short_name)
```
Instead of trying to come up with a good path here, I think it would be much better to simply provide the name of the output file (`variable['filename']`) to the function `fix_file` in namelist.py and write the fixed file to that location.

jvegasbsc replied:
We usually have more than on file at this step

jvegasbsc replied:
Ok. I will use msktemp with the variable[filename] as a prefix

bouweandela commented on esmvaltool/preprocessor/_reformat.py:
```diff
@@ -51,7 +57,8 @@ def fix_metadata(cube, short_name, project, model, cmor_table=None, mip=None):
 
 
-def fix_data(cube, short_name, project, model, cmor_table=None, mip=None):
+def fix_data(cube, short_name, project, model, preproc_dir,
```
The new code here seems unrelated to fixing the data in the cube, but it looks like it is trying to clean up after the function `fix_file`. This should probably be removed.

jvegasbsc replied:
Yes, to avoid having the fixed files forever in the folder. At which step do you suggest that we clean them?

bouweandela replied:
I think it's probably best to create a separate `cleanup` preprocessor function, which we run at the very end, instead of adding it to some unrelated preprocessor function.

jvegasbsc replied:
Ok, I will add it

jvegasbsc commented:
There are only two: 
* iris.exceptions can not be loaded -> I don't know why this is failing, the import is ok
* Too many arguments (7 / 5) in fix_data -> I added a parameter here (discussing with @bouweandela where to put this code now) and interpreted this as if I fixed "Too many arguments (6/ 5) in fix_data" and added this...

bouweandela commented on esmvaltool/namelist.py:
```diff
@@ -288,16 +288,16 @@ def _get_default_settings(variable, config_user):
         'model': variable['model'],
         'short_name': variable['short_name'],
-        'preproc_dir': config_user['preproc_dir'],
+        'var_path': variable['var_path'],
```
You're right that there can be multiple fixed output files, I didn't think about that. How about replacing the line above by a more descriptive:
```
'output_dir': variable['filename'] + '_fixed'
```
and then creating any fixed files in that directory. The same directory can then be provided to the new `cleanup` function, e.g. 
```
settings['cleanup'] = {
    'remove':  variable['filename'] + '_fixed',
}
````
to be deleted at the very end of the preprocessor run.

jvegasbsc replied:
Nice and cleaner

jvegasbsc replied:
If save_intermediate_files is True, we should skip the cleanup, no?

jvegasbsc commented:
Wait, it is not working!

jvegasbsc commented:
A couple of comments:
 - I put the cleanup function before saving on purpose: most of the times storage quota is an issue on HPCs and the original fixed files size can be huge. We should clean them as soon as we are sure that all the data is in memory

- I created a special category for tasks that do not use and produce data. I think it is cleaner to do this that a task that gets a file list, do some things without using it and returns the same file list

bouweandela commented:
My apologies for making a mess by directly committing to the branch, I should have discussed first. 
- The reason that I think the cleanup should be done after saving, is that iris only loads data when you actually access it, see [here](http://scitools.org.uk/iris/docs/latest/userguide/loading_iris_cubes.html#lazy-loading). Therefore I think it is possible that data will only be loaded into memory at the time you save to file.
- The no data case doesn't follow 'pipeline' pattern that is the basis of how the preprocessor is implemented. If there would be many preprocessor functions that do not require data, I think it could make sense to specify a special no data case, but since there is only one now and I do not foresee adding more, I think it adds unnessary complication.

jvegasbsc commented:
> My apologies for making a mess by directly committing to the branch, I should have discussed first.
I missed your commit, I found an issue with my own version...

>The reason that I think the cleanup should be done after saving, is that iris only loads data when you actually access it, see here. Therefore I think it is possible that data will only be loaded into memory at the time you save to file.

Yes, you are right. I thought that it would be impossible to achieve that because of the check_data step, but if a variable does not have valid_min or valid_max values is entirely possible

>The no data case doesn't follow 'pipeline' pattern that is the basis of how the preprocessor is implemented. If there would be many preprocessor functions that do not require data, I think it could make sense to specify a special no data case, but since there is only one now and I do not foresee adding more, I think it adds unnessary complication.

Ok, I will revert that change

bouweandela commented:
It would be good if you could add `'cleanup'` to `esmvaltool.preprocessor._LIST_INPUT_FUNCTIONS` and `esmvaltool.preprocessor._LIST_OUTPUT_FUNCTIONS`, like I did in this commit eba9532062c78954d6456b56bf2e8bb107193ca5, that way it will only be run once for every output file.

jvegasbsc commented:
Fix done.

mattiarighi commented:
I get an error when running `namelist_perfmetrics_CMIP5.yml` (short version as well):

```
  File "/public/esmvaltool/namelist.py", line 133, in check_preprocessor_settings
    "function {}".format(invalid_args, step))
NamelistError: Invalid argument(s) set(['output_dir']) encountered for preprocessor function fix_file
```
